housekeeping/src/main/groovy/whelk/housekeeping/NotificationSender.groovy

```diff
@@ -0,0 +1,429 @@
+        Map<String, List<Map>> heldByToUserSettings = new HashMap<>();
+        {
+            List<Map> allUserSettingStrings = whelk.getStorage().getAllUserData()
+            for (Map settings : allUserSettingStrings) {
+                if (!settings["notificationEmail"])
+                    continue
+                settings?.requestedNotifications.each { request ->
+                    if (!request instanceof Map)
+                        return
+                    if (!request["heldBy"])
+                        return
+
+                    String heldBy = request["heldBy"]
+                    if (!heldByToUserSettings.containsKey(heldBy))
+                        heldByToUserSettings.put(heldBy, [])
+                    List userSettingsForThisHeldBy = heldByToUserSettings[heldBy]
+                    userSettingsForThisHeldBy.add(settings)
+                }
+            }
+        }
+
+        // Determine the time interval of changes for which to generate notifications.
+        Timestamp from = Timestamp.from(Instant.now().minus(1, ChronoUnit.DAYS)) // Default to last 24h if first time.
+        Map state = whelk.getStorage().getState(STATE_KEY)
+        if (state && state.lastEmailTime)
+            from = Timestamp.from( ZonedDateTime.parse( (String) state.lastEmailTime, DateTimeFormatter.ISO_OFFSET_DATE_TIME).toInstant() )
+        Timestamp until = Timestamp.from(Instant.now())
+
+        Connection connection
+        PreparedStatement statement
+        ResultSet resultSet
+
+        connection = whelk.getStorage().getOuterConnection()
+        connection.setAutoCommit(false)
+        try {
+            // Fetch all changed IDs within the interval
+            String sql = "SELECT id FROM lddb WHERE collection IN ('bib', 'auth') AND ( modified > ? AND modified <= ? );"
+            connection.setAutoCommit(false)
+            statement = connection.prepareStatement(sql)
+            statement.setTimestamp(1, from)
+            statement.setTimestamp(2, until)
+            statement.setFetchSize(512)
+            resultSet = statement.executeQuery()
+            // If both an instance and one of it's dependencies are affected within the same interval, we will
+            // (without this check) try to generate notifications for said instance twice.
+            Set affectedInstanceIDs = []
+            Map<String, List<String>> notificationsByUser = new HashMap<>()
+            while (resultSet.next()) {
+                String id = resultSet.getString("id")
+                generateNotificationsForChangedID(id, heldByToUserSettings, from.toInstant(),
+                        until.toInstant(), affectedInstanceIDs, notificationsByUser)
+            }
+
+            notificationsByUser.keySet().each { email ->
+                List<String> notifications = notificationsByUser.get(email)
+                sendEmail(senderAddress, email, "Förändingsmeddelande", notifications.join("\n"))
```

Comment:
```suggestion
                sendEmail(senderAddress, email, "Förändringsmeddelande", notifications.join("\n"))
```
Hard coding the subject feels a bit icky. Using java.util.Locale (+ java.util.ResourceBundle) is - maybe? - a bit overkill for just a few lines given that we (AFAIK!) don't have a need for it anywhere else. So, it's just a few lines in total, maybe put them in secret.properties? :shrug:

Reply:
I agree, it feels icky!
But ResourceBundle (or secret.properties) would be far worse.
I still think this is the lesser evil.

Reply:
Can there be no middleground; e.g. an `se.kb.libris`-namespaced class with constants?